cps/kni.c

```diff
@@ -924,6 +924,11 @@ rd_send_batch(struct cps_config *cps_conf, struct mnl_nlmsg_batch *batch,
 		done->nlmsg_flags = NLM_F_MULTI;
 		done->nlmsg_seq = seq;
 		done->nlmsg_pid = cps_conf->nl_pid;
+		if (!mnl_nlmsg_batch_next(batch)) {
+			ret = -ENOMEM;
+			CPS_LOG(ERR, "Can't add Netlink NLMSG_DONE to end of batch\n");
+			goto reset;
```

Comment:
Do we really need to deal this case as an error since the DONE message is already in the batch? I think we should ignore the result of `mnl_nlmsg_batch_next()` in this case and have a comment explaining why it's important to call it anyway.

Reply:
`mnl_nlmsg_batch_next()` returns `false` if the last message (the DONE message) did not fit into the batch (the Netlink packet's size has gone over `MNL_SOCKET_BUFFER_SIZE`). This would leave the interaction unterminated if the requester did not specify `NLM_F_ACK`, but we can end it with a (separate) ERROR message to allow the requester to continue.

Reply:
My suggestion is to call `rd_send_batch()` with parameter `done = false` as it's done in `rd_getroute_ipv4_locked()`.

Reply:
If `mnl-nlmsg_batch_next()` fails, it doesn't update the size of the batch, so if we call `rd_send_batch()` (recursively?) with `done = false` at this point the message will be sent but it still won't contain the DONE message. So we will still need to send a message of type `NLMSG_ERROR` back to the daemon (either as an ack or with an error code) in order to terminate the interaction.

Reply:
I think we are interpreting things differently.

In my mind, calling `rd_send_batch()` with `done = false` will send the batch out without the DONE message and reset the batch; what it's fine because we are not done. Only the `rd_send_batch()` at the end of this function will call `rd_send_batch()` with `done = true`.

Reply:
I see -- so something like send the batch without the DONE, reset the batch, then just send the DONE using the same batch buffer?

Reply:
Yes. This is what we do in `rd_getroute_ipv4_locked()`.

Reply:
Makes sense now -- will do.